pkg/resources/reconciling/modifier/related_revisions.go

```diff
@@ -0,0 +1,217 @@
+/*
+Copyright 2024 The Kubermatic Kubernetes Platform contributors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modifier
+
+import (
+	"context"
+	"fmt"
+
+	"k8c.io/reconciler/pkg/reconciling"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// RelatedRevisionsLabels scans volume mounts for pod templates for ConfigMaps
+// and Secrets and will then put new labels for these mounts onto the pod template, causing
+// restarts when the volumes changed.
+func RelatedRevisionsLabels(ctx context.Context, client ctrlruntimeclient.Client) reconciling.ObjectModifier {
+	return func(reconciler reconciling.ObjectReconciler) reconciling.ObjectReconciler {
+		return func(existing ctrlruntimeclient.Object) (ctrlruntimeclient.Object, error) {
+			obj, err := reconciler(existing)
+			if err != nil {
+				return obj, err
+			}
+
+			return addRevisionLabelsToObject(ctx, client, obj)
+		}
+	}
+}
+
+func addRevisionLabelsToObject(ctx context.Context, client ctrlruntimeclient.Client, obj ctrlruntimeclient.Object) (ctrlruntimeclient.Object, error) {
+	switch asserted := obj.(type) {
+	case *appsv1.Deployment:
+		if err := addRevisionLabelsToPodSpec(ctx, client, asserted.Namespace, &asserted.Spec.Template); err != nil {
+			return nil, fmt.Errorf("failed to determine revision labels: %w", err)
+		}
+
+	case *appsv1.StatefulSet:
+		if err := addRevisionLabelsToPodSpec(ctx, client, asserted.Namespace, &asserted.Spec.Template); err != nil {
+			return nil, fmt.Errorf("failed to determine revision labels: %w", err)
+		}
+
+	case *appsv1.DaemonSet:
+		if err := addRevisionLabelsToPodSpec(ctx, client, asserted.Namespace, &asserted.Spec.Template); err != nil {
+			return nil, fmt.Errorf("failed to determine revision labels: %w", err)
+		}
+	}
+
+	return obj, nil
```

Comment:
Why we don't have a default case that's going to throw an error if the object is not one of these three types? We have something like that in the version label modifier. My worry here is that someone might for any reason try to use this modifier on some other object type, and that would do nothing, but they might not expect that.

Reply:
Agreed, I added a panic().